services/horizon/internal/ingest/fsm_history_range_state.go

```diff
@@ -0,0 +1,107 @@
+package ingest
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/stellar/go/support/errors"
+	logpkg "github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+)
+
+type historyRangeState struct {
+	fromLedger uint32
+	toLedger   uint32
+}
+
+func (h historyRangeState) String() string {
+	return fmt.Sprintf(
+		"historyRange(fromLedger=%d, toLedger=%d)",
+		h.fromLedger,
+		h.toLedger,
+	)
+}
+
+func (historyRangeState) GetState() State {
+	return HistoryRange
+}
+
+// historyRangeState is used when catching up history data
+func (h historyRangeState) run(s *system) (transition, error) {
+	if h.fromLedger == 0 || h.toLedger == 0 ||
+		h.fromLedger > h.toLedger {
+		return start(), errors.Errorf("invalid range: [%d, %d]", h.fromLedger, h.toLedger)
+	}
+
+	err := s.maybePrepareRange(s.ctx, h.fromLedger)
+	if err != nil {
+		return start(), err
+	}
+
+	if err = s.historyQ.Begin(s.ctx); err != nil {
+		return start(), errors.Wrap(err, "Error starting a transaction")
+	}
+	defer s.historyQ.Rollback()
+
+	// acquire distributed lock so no one else can perform ingestion operations.
+	if _, err = s.historyQ.GetLastLedgerIngest(s.ctx); err != nil {
+		return start(), errors.Wrap(err, getLastIngestedErrMsg)
+	}
+
+	lastHistoryLedger, err := s.historyQ.GetLatestHistoryLedger(s.ctx)
+	if err != nil {
+		return start(), errors.Wrap(err, "could not get latest history ledger")
+	}
+
+	// We should be ingesting the ledger which occurs after
+	// lastHistoryLedger. Otherwise, some other horizon node has
+	// already completed the ingest history range operation and
+	// we should go back to the init state
+	if lastHistoryLedger != h.fromLedger-1 {
+		return start(), nil
+	}
+
+	ledgers := []xdr.LedgerCloseMeta{}
```

Comment:
```suggestion
	ledgers := make([]xdr.LedgerCloseMeta, 0, s.maxLedgerPerFlush)
```

you can reset the slice after flushing with this code:

```go
ledgers = ledgers[0:]
```

Reply:
thanks for the suggestions on better slice handling, worked that in - https://github.com/stellar/go/pull/5117/commits/c2aea6ab64612b13b934c446ed0b92fc7f5afa30